drivers/audio/wm8904.c

```diff
@@ -0,0 +1,728 @@
+	const uint16_t val = WM8904_REGVAL_OUT_VOL(0, 0, 1, volume);
+	const uint16_t mask = WM8904_REGMASK_OUT_VU
+		| WM8904_REGMASK_OUT_ZC
+		| WM8904_REGMASK_OUT_VOL;
+
+	switch (channel) {
+	case AUDIO_CHANNEL_FRONT_LEFT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT2_LEFT, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_FRONT_RIGHT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT2_RIGHT, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_HEADPHONE_LEFT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_LEFT, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_HEADPHONE_RIGHT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_RIGHT, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_ALL:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_LEFT, mask, val);
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_RIGHT, mask, val);
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT2_LEFT, mask, val);
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT2_RIGHT, mask, val);
+		return 0;
+
+	default:
+		return -EINVAL;
+	}
+}
+
+static int wm8904_out_mute_config(const struct device *dev, audio_channel_t channel, bool mute)
+{
+	const uint16_t val = WM8904_REGVAL_OUT_VOL(mute, 0, 0, 0);
+	const uint16_t mask = WM8904_REGMASK_OUT_MUTE;
+
+	switch (channel) {
+	case AUDIO_CHANNEL_FRONT_LEFT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT2_LEFT, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_FRONT_RIGHT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT2_RIGHT, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_HEADPHONE_LEFT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_LEFT, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_HEADPHONE_RIGHT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_RIGHT, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_ALL:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_LEFT, mask, val);
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_RIGHT, mask, val);
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT2_LEFT, mask, val);
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_RIGHT, mask, val);
+		return 0;
+
+	default:
+		return -EINVAL;
+	}
+}
+
+static int wm8904_in_volume_config(const struct device *dev, audio_channel_t channel, int volume)
+{
+	const uint16_t val = WM8904_REGVAL_IN_VOL(0, volume);
+	const uint16_t mask = WM8904_REGMASK_IN_MUTE;
+
+	switch (channel) {
+	case AUDIO_CHANNEL_FRONT_LEFT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_LEFT_IN_0, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_FRONT_RIGHT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_RIGHT_IN_0, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_ALL:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_LEFT_IN_0, mask, val);
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_RIGHT_IN_0, mask, val);
+		return 0;
+
+	default:
+		return -EINVAL;
+	}
+}
+
+static int wm8904_in_mute_config(const struct device *dev, audio_channel_t channel, bool mute)
+{
+	const uint16_t val = WM8904_REGVAL_IN_VOL(mute, 0);
+	const uint16_t mask = WM8904_REGMASK_IN_MUTE;
+
+	switch (channel) {
```

Comment:
Similarly here. It is the same as in `wm8904_in_mute_config()`.

Reply:
@anangl Personally I think that the code can be followed a bit more easily than if one strives to avoid even two-fold repetitions at all cost, but if you insist...